File: components/monitoring/sdk/c/src/runtime_config.c
/**
 * @file runtime_config.c
 * @brief Runtime Configuration Implementation
 */

#include "monitoring_runtime_config.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <unistd.h>
#include <pthread.h>

/* JSON parsing - using a simple parser for demonstration */
/* In production, use a library like cJSON or json-c */

typedef struct {
    monitoring_runtime_config_options_t options;
    monitoring_config_t default_config;
    monitoring_config_t current_config;
    
    pthread_t watcher_thread;
    bool watcher_running;
    pthread_mutex_t config_lock;
    
    time_t last_modified;
    time_t last_reload_time;
    bool last_reload_success;
    
    bool initialized;
} runtime_config_state_t;

static runtime_config_state_t g_runtime_state = {
    .initialized = false,
    .watcher_running = false,
    .last_modified = 0,
    .last_reload_time = 0,
    .last_reload_success = false
};

/* Forward declarations */
static void* config_watcher_thread(void* arg);
static monitoring_error_t load_config_from_file(const char* path, monitoring_config_t* config);
static time_t get_file_mtime(const char* path);
static monitoring_error_t apply_new_config(const monitoring_config_t* new_config);

monitoring_error_t monitoring_init_with_runtime_config(
    const monitoring_config_t* default_config,
    const monitoring_runtime_config_options_t* runtime_options)
{
    if (!default_config || !runtime_options) {
        return MONITORING_INVALID_PARAM;
    }
    
    if (g_runtime_state.initialized) {
        return MONITORING_ALREADY_INIT;
    }
    
    /* Store default configuration */
    memcpy(&g_runtime_state.default_config, default_config, sizeof(monitoring_config_t));
    memcpy(&g_runtime_state.options, runtime_options, sizeof(monitoring_runtime_config_options_t));
    
    pthread_mutex_init(&g_runtime_state.config_lock, NULL);
    
    /* Try to load config from file */
    monitoring_config_t loaded_config;
    monitoring_error_t err = load_config_from_file(
        runtime_options->config_file_path,
        &loaded_config
    );
    
    if (err == MONITORING_OK) {
        /* Use loaded config */
        memcpy(&g_runtime_state.current_config, &loaded_config, sizeof(monitoring_config_t));
        fprintf(stderr, "[runtime_config] Loaded config from: %s\n", 
                runtime_options->config_file_path);
    } else if (runtime_options->use_fallback) {
        /* Use default config as fallback */
        memcpy(&g_runtime_state.current_config, default_config, sizeof(monitoring_config_t));
        fprintf(stderr, "[runtime_config] Using default config (file load failed: %s)\n",
                monitoring_error_string(err));
    } else {
        return err;
    }
    
    /* Initialize SDK with current config */
    err = monitoring_init(&g_runtime_state.current_config);
    if (err != MONITORING_OK) {
        pthread_mutex_destroy(&g_runtime_state.config_lock);
        return err;
    }
    
    g_runtime_state.last_modified = get_file_mtime(runtime_options->config_file_path);
    g_runtime_state.initialized = true;
    
    /* Start config watcher thread if auto-reload enabled */
    if (runtime_options->auto_reload) {
        g_runtime_state.watcher_running = true;
        pthread_create(&g_runtime_state.watcher_thread, NULL, 
                      config_watcher_thread, NULL);
    }
    
    return MONITORING_OK;
}

monitoring_error_t monitoring_reload_config(void)
{
    if (!g_runtime_state.initialized) {
        return MONITORING_NOT_INITIALIZED;
    }
    
    pthread_mutex_lock(&g_runtime_state.config_lock);
    
    monitoring_config_t new_config;
    monitoring_error_t err = load_config_from_file(
        g_runtime_state.options.config_file_path,
        &new_config
    );
    
    if (err == MONITORING_OK) {
        /* Apply new configuration */
        err = apply_new_config(&new_config);
        
        if (err == MONITORING_OK) {
            memcpy(&g_runtime_state.current_config, &new_config, 
                   sizeof(monitoring_config_t));
            g_runtime_state.last_reload_time = time(NULL);
            g_runtime_state.last_reload_success = true;
            
            fprintf(stderr, "[runtime_config] Configuration reloaded successfully\n");
            
            /* Trigger callback */
            if (g_runtime_state.options.on_config_reload) {
                g_runtime_state.options.on_config_reload(true, "Configuration reloaded");
            }
        } else {
            g_runtime_state.last_reload_success = false;
            
            fprintf(stderr, "[runtime_config] Failed to apply new config: %s\n",
                   monitoring_error_string(err));
            
            if (g_runtime_state.options.on_config_reload) {
                g_runtime_state.options.on_config_reload(false, monitoring_error_string(err));
            }
        }
    } else {
        g_runtime_state.last_reload_success = false;
        
        fprintf(stderr, "[runtime_config] Failed to load config file: %s\n",
               monitoring_error_string(err));
        
        if (g_runtime_state.options.on_config_reload) {
            g_runtime_state.options.on_config_reload(false, monitoring_error_string(err));
        }
    }
    
    pthread_mutex_unlock(&g_runtime_state.config_lock);
    
    return err;
}

const char* monitoring_get_config_file_path(void)
{
    if (!g_runtime_state.initialized) {
        return NULL;
    }
    return g_runtime_state.options.config_file_path;
}

monitoring_error_t monitoring_set_auto_reload(bool enabled)
{
    if (!g_runtime_state.initialized) {
        return MONITORING_NOT_INITIALIZED;
    }
    
    pthread_mutex_lock(&g_runtime_state.config_lock);
    
    if (enabled && !g_runtime_state.watcher_running) {
        /* Start watcher */
        g_runtime_state.watcher_running = true;
        pthread_create(&g_runtime_state.watcher_thread, NULL,
                      config_watcher_thread, NULL);
    } else if (!enabled && g_runtime_state.watcher_running) {
        /* Stop watcher */
        g_runtime_state.watcher_running = false;
        pthread_join(g_runtime_state.watcher_thread, NULL);
    }
    
    g_runtime_state.options.auto_reload = enabled;
    
    pthread_mutex_unlock(&g_runtime_state.config_lock);
    
    return MONITORING_OK;
}

monitoring_error_t monitoring_get_reload_status(time_t* timestamp, bool* success)
{
    if (!g_runtime_state.initialized) {
        return MONITORING_NOT_INITIALIZED;
    }
    
    pthread_mutex_lock(&g_runtime_state.config_lock);
    
    if (timestamp) {
        *timestamp = g_runtime_state.last_reload_time;
    }
    
    if (success) {
        *success = g_runtime_state.last_reload_success;
    }
    
    pthread_mutex_unlock(&g_runtime_state.config_lock);
    
    return MONITORING_OK;
}

/* ========================================================================== */
/* Internal Functions                                                         */
/* ========================================================================== */

static void* config_watcher_thread(void* arg)
{
    (void)arg;
    
    while (g_runtime_state.watcher_running) {
        sleep(g_runtime_state.options.check_interval_seconds);
        
        if (!g_runtime_state.watcher_running) {
            break;
        }
        
        /* Check if config file has been modified */
        time_t current_mtime = get_file_mtime(g_runtime_state.options.config_file_path);
        
        if (current_mtime > g_runtime_state.last_modified) {
            fprintf(stderr, "[runtime_config] Config file changed, reloading...\n");
            
            /* Small delay to ensure file write is complete */
            usleep(100000); /* 100ms */
            
            monitoring_reload_config();
            
            g_runtime_state.last_modified = current_mtime;
        }
    }
    
    return NULL;
}

static time_t get_file_mtime(const char* path)
{
    struct stat st;
    if (stat(path, &st) == 0) {
        return st.st_mtime;
    }
    return 0;
}

static monitoring_error_t load_config_from_file(const char* path, monitoring_config_t* config)
{
    if (!path || !config) {
        return MONITORING_INVALID_PARAM;
    }
    
    FILE* fp = fopen(path, "r");
    if (!fp) {
        return MONITORING_IO_ERROR;
    }
    
    /* Read file content */
    fseek(fp, 0, SEEK_END);
    long size = ftell(fp);
    fseek(fp, 0, SEEK_SET);
    
    char* content = malloc(size + 1);
    if (!content) {
        fclose(fp);
        return MONITORING_NO_MEMORY;
    }
    
    fread(content, 1, size, fp);
    content[size] = '\0';
    fclose(fp);
    
    /* Simple JSON parsing (in production, use a proper JSON library) */
    /* For now, we'll implement a minimal parser for demonstration */
    
    /* Extract mode */
    char* mode_str = strstr(content, "\"mode\"");
    if (mode_str) {
        if (strstr(mode_str, "\"sidecar\"")) {
            config->mode = MONITORING_MODE_SIDECAR;
        } else if (strstr(mode_str, "\"direct\"")) {
            config->mode = MONITORING_MODE_DIRECT;
        }
    }
    
    /* Extract app info */
    char* app_name = strstr(content, "\"name\"");
    if (app_name) {
        /* Parse app name (simplified) */
        char* start = strchr(app_name + 6, '"');
        if (start) {
            start++;
            char* end = strchr(start, '"');
            if (end) {
                size_t len = end - start;
                if (len < 256) {
                    strncpy((char*)config->app_name, start, len);
                    ((char*)config->app_name)[len] = '\0';
                }
            }
        }
    }
    
    /* Note: Full JSON parsing would extract all backends, configurations, etc. */
    /* This is a simplified demonstration. Use cJSON or similar in production. */
    
    free(content);
    
    return MONITORING_OK;
}

static monitoring_error_t apply_new_config(const monitoring_config_t* new_config)
{
    /* In a full implementation, this would:
     * 1. Initialize new backends
     * 2. Keep old backends active during transition
     * 3. Gracefully shutdown old backends
     * 4. Swap to new backends atomically
     * 
     * For now, we'll do a simplified version
     */
    
    /* TODO: Implement hot-swap logic */
    /* This requires backend interface changes to support:
     * - Backend lifecycle management
     * - Graceful shutdown
     * - Atomic swapping
     */
    
    return MONITORING_OK;
}

File: components/monitoring/sdk/r/R/runtime_config.R
#' Runtime Configuration with Hot-Reloading for R SDK
#'
#' @description
#' Enables dynamic backend routing changes without restarting R sessions.
#'
#' @section Features:
#' \itemize{
#'   \item{Load configuration from YAML files}
#'   \item{Auto-detection of file changes}
#'   \item{Hot-reload backends without restart}
#'   \item{Callback support for reload events}
#'   \item{Fallback to default configuration}
#' }
#'
#' @name runtime_config
NULL

# Runtime config state (module-level)
.runtime_state <- new.env(parent = emptyenv())
.runtime_state$initialized <- FALSE
.runtime_state$config_file <- NULL
.runtime_state$check_interval <- 30
.runtime_state$auto_reload <- FALSE
.runtime_state$on_reload <- NULL
.runtime_state$last_mtime <- 0
.runtime_state$last_reload_time <- 0
.runtime_state$last_reload_success <- FALSE
.runtime_state$timer_id <- NULL

#' Initialize SDK with Runtime Configuration
#'
#' @param config_file Path to YAML configuration file
#' @param default_config List with default configuration (fallback)
#' @param auto_reload Logical. Enable automatic config reloading
#' @param check_interval Numeric. Check interval in seconds (default: 30)
#' @param on_reload Function(success, message) callback for reload events
#' @param use_fallback Logical. Use default config if file not found
#'
#' @return Invisible NULL
#' @export
#'
#' @examples
#' \dontrun{
#' default_config <- list(
#'   mode = "sidecar",
#'   app = list(name = "my-r-app", version = "1.0.0", site_id = "fab1"),
#'   sidecar = list(url = "http://localhost:17000")
#' )
#' 
#' monitoring_init_with_runtime_config(
#'   config_file = "monitoring_config.yaml",
#'   default_config = default_config,
#'   auto_reload = TRUE,
#'   on_reload = function(success, message) {
#'     if (success) message("✓ Config reloaded: ", message)
#'     else warning("✗ Config reload failed: ", message)
#'   }
#' )
#' }
monitoring_init_with_runtime_config <- function(
  config_file,
  default_config = NULL,
  auto_reload = TRUE,
  check_interval = 30,
  on_reload = NULL,
  use_fallback = TRUE
) {
  if (.runtime_state$initialized) {
    stop("Runtime config already initialized. Call monitoring_shutdown_runtime_config() first.")
  }
  
  # Store runtime options
  .runtime_state$config_file <- config_file
  .runtime_state$check_interval <- check_interval
  .runtime_state$auto_reload <- auto_reload
  .runtime_state$on_reload <- on_reload
  
  # Try to load config from file
  config <- .load_config_from_file(config_file)
  
  if (is.null(config) && use_fallback && !is.null(default_config)) {
    config <- default_config
    message("Using default config as fallback")
  } else if (is.null(config)) {
    stop("Failed to load config and no fallback available")
  }
  
  # Initialize SDK with loaded config
  tryCatch({
    do.call(monitoring_init, config)
    .runtime_state$last_reload_time <- Sys.time()
    .runtime_state$last_reload_success <- TRUE
    message("✓ SDK initialized with runtime config from: ", config_file)
  }, error = function(e) {
    if (!is.null(on_reload)) {
      on_reload(FALSE, paste("Init failed:", e$message))
    }
    stop(e)
  })
  
  # Get initial file mtime
  if (file.exists(config_file)) {
    .runtime_state$last_mtime <- file.info(config_file)$mtime
  }
  
  # Start auto-reload timer if enabled
  if (auto_reload) {
    .start_file_watcher()
  }
  
  .runtime_state$initialized <- TRUE
  
  if (!is.null(on_reload)) {
    on_reload(TRUE, "Initial configuration loaded")
  }
  
  invisible(NULL)
}

#' Manually Reload Configuration
#'
#' @return Logical. TRUE if reload successful, FALSE otherwise
#' @export
monitoring_reload_runtime_config <- function() {
  if (!.runtime_state$initialized) {
    warning("Runtime config not initialized")
    return(FALSE)
  }
  
  message("Manually reloading configuration...")
  return(.reload_config())
}

#' Get Runtime Configuration Status
#'
#' @return List with reload status information
#' @export
monitoring_get_runtime_status <- function() {
  if (!.runtime_state$initialized) {
    return(list(initialized = FALSE))
  }
  
  list(
    initialized = .runtime_state$initialized,
    config_file = .runtime_state$config_file,
    auto_reload = .runtime_state$auto_reload,
    check_interval = .runtime_state$check_interval,
    last_reload_time = .runtime_state$last_reload_time,
    last_reload_success = .runtime_state$last_reload_success
  )
}

#' Enable or Disable Auto-Reload
#'
#' @param enabled Logical. TRUE to enable, FALSE to disable
#' @export
monitoring_set_auto_reload <- function(enabled) {
  if (!.runtime_state$initialized) {
    warning("Runtime config not initialized")
    return(invisible(NULL))
  }
  
  if (enabled && !.runtime_state$auto_reload) {
    .start_file_watcher()
    .runtime_state$auto_reload <- TRUE
    message("Auto-reload enabled")
  } else if (!enabled && .runtime_state$auto_reload) {
    .stop_file_watcher()
    .runtime_state$auto_reload <- FALSE
    message("Auto-reload disabled")
  }
  
  invisible(NULL)
}

#' Shutdown Runtime Configuration
#'
#' @export
monitoring_shutdown_runtime_config <- function() {
  if (!.runtime_state$initialized) {
    return(invisible(NULL))
  }
  
  .stop_file_watcher()
  monitoring_shutdown()
  
  .runtime_state$initialized <- FALSE
  .runtime_state$config_file <- NULL
  
  message("✓ Runtime config shut down")
  invisible(NULL)
}

# ============================================================================
# Internal Functions
# ============================================================================

.load_config_from_file <- function(file_path) {
  if (!file.exists(file_path)) {
    warning("Config file not found: ", file_path)
    return(NULL)
  }
  
  tryCatch({
    # Load YAML config
    if (!requireNamespace("yaml", quietly = TRUE)) {
      stop("yaml package required for config files. Install with: install.packages('yaml')")
    }
    
    config <- yaml::read_yaml(file_path)
    message("Config loaded from: ", file_path)
    return(config)
    
  }, error = function(e) {
    warning("Failed to load config file: ", e$message)
    return(NULL)
  })
}

.reload_config <- function() {
  config <- .load_config_from_file(.runtime_state$config_file)
  
  if (is.null(config)) {
    .runtime_state$last_reload_success <- FALSE
    .runtime_state$last_reload_time <- Sys.time()
    
    msg <- "Failed to load config file"
    warning(msg)
    if (!is.null(.runtime_state$on_reload)) {
      .runtime_state$on_reload(FALSE, msg)
    }
    return(FALSE)
  }
  
  tryCatch({
    # Shutdown current SDK
    monitoring_shutdown()
    
    # Re-initialize with new config
    do.call(monitoring_init, config)
    
    .runtime_state$last_reload_success <- TRUE
    .runtime_state$last_reload_time <- Sys.time()
    
    msg <- "Configuration reloaded successfully"
    message("✓ ", msg)
    if (!is.null(.runtime_state$on_reload)) {
      .runtime_state$on_reload(TRUE, msg)
    }
    
    return(TRUE)
    
  }, error = function(e) {
    .runtime_state$last_reload_success <- FALSE
    .runtime_state$last_reload_time <- Sys.time()
    
    msg <- paste("Failed to apply new config:", e$message)
    warning(msg)
    if (!is.null(.runtime_state$on_reload)) {
      .runtime_state$on_reload(FALSE, msg)
    }
    
    return(FALSE)
  })
}

.check_file_changed <- function() {
  if (!file.exists(.runtime_state$config_file)) {
    return(FALSE)
  }
  
  current_mtime <- file.info(.runtime_state$config_file)$mtime
  
  if (current_mtime > .runtime_state$last_mtime) {
    # File was modified
    .runtime_state$last_mtime <- current_mtime
    
    # Small delay to ensure file write is complete
    Sys.sleep(0.1)
    
    message("Config file changed, reloading...")
    .reload_config()
    
    return(TRUE)
  }
  
  return(FALSE)
}

.file_watcher_task <- function() {
  .check_file_changed()
}

.start_file_watcher <- function() {
  if (!is.null(.runtime_state$timer_id)) {
    return()  # Already running
  }
  
  # Schedule periodic check
  # Note: Using later package for async task scheduling
  if (!requireNamespace("later", quietly = TRUE)) {
    warning("later package recommended for auto-reload. Install with: install.packages('later')")
    warning("Auto-reload disabled. Use manual reload instead.")
    return()
  }
  
  # Create recurring task
  .runtime_state$timer_id <- later::later(
    function() {
      .file_watcher_task()
      # Reschedule
      if (.runtime_state$auto_reload) {
        .start_file_watcher()
      }
    },
    delay = .runtime_state$check_interval
  )
  
  message("File watcher started (check interval: ", .runtime_state$check_interval, "s)")
}

.stop_file_watcher <- function() {
  # In R, we can't easily cancel a later() task once scheduled
  # Just set auto_reload to FALSE to prevent rescheduling
  .runtime_state$timer_id <- NULL
}

